### PR TITLE
[BB-3512] Added support for course discovery bucket configuration

### PR DESCRIPTION
**Pull request by farhaanbukhsh**
This PR helps to create a S3 bucket for course-discovery ussage. This helps to create the IAM user
as well as the policy.

**JIRA tickets**: [BB-3512](https://tasks.opencraft.com/browse/BB-3512)

~~**Discussions**: Link to any public dicussions about this PR or the design/architecture. Otherwise omit this.~~

~~**Dependencies**: None~~

~~**Screenshots**: Always include screenshots if there is any change to the UI.~~

~~**Sandbox URL**: TBD - sandbox is being provisioned.~~

~~**Merge deadline**: "None" if there's no rush, "ASAP" if it's critical, or provide a specific date if there is one.~~

**Testing instructions**:
TBD

**Reviewers**
- [ ] @giovannicimolin 
- [ ] @guruprasad

### CODE REVIEW

File: modules/openedx/application/outputs.tf [hunk 1]
```diff
@@ -37,3 +37,7 @@ output "rds" {
 output "redis" {
   value = aws_elasticache_cluster.redis
 }
+
+output "edxapp_course_discovery_s3_access_key" {
```
Comment (giovannicimolin) [inline modules/openedx/application/outputs.tf]: @farhaanbukhsh This output should be on the `s3/outputs.tf` file (they error out here).
There you can add:
```
output "edxapp_course_discovery_s3_access_key" {
  value = aws_iam_access_key.edxapp_course_discovery_s3_access_key
}
```

File: modules/openedx/s3/course_discovery.tf [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      aws_s3_bucket.edxapp_course_discovery.arn,
+      "${aws_s3_bucket.edxapp_course_discovery.arn}/*",
+    ]
+  }
+}
+
+
+resource aws_s3_bucket edxapp_course_discovery {
+  count = var.enable_course_discovery ? 1 : 0
+  bucket = "${var.customer_name}-${var.environment}-edxapp-course-discovery"
```
Comment (giovannicimolin) [inline modules/openedx/s3/course_discovery.tf]: It's not `var.customer_name` here (see https://github.com/open-craft/terraform-scripts/blob/main/modules/openedx/application/s3.tf#L8):
```suggestion
  bucket = "${var.client_shortname}-${var.environment}-edxapp-course-discovery"
```
Replace all occurrences in this file.

File: modules/openedx/s3/course_discovery.tf [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  version = "2012-10-17"
+  statement {
+    actions = [
+      "s3:*",
+    ]
+    effect = "Allow"
+
+    resources = [
+      aws_s3_bucket.edxapp_course_discovery.arn,
+      "${aws_s3_bucket.edxapp_course_discovery.arn}/*",
```
Comment (giovannicimolin) [inline modules/openedx/s3/course_discovery.tf]: When using count, you'll need to add an index to the element you're fetching:
```suggestion
      aws_s3_bucket.edxapp_course_discovery[0].arn,
      "${aws_s3_bucket.edxapp_course_discovery[0].arn}/*",
```

File: modules/openedx/s3/course_discovery.tf [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    allowed_origins = ["*"]
+  }
+}
+
+resource "aws_iam_policy" "edxapp_course_discovery_s3_full_access" {
+  count = var.enable_course_discovery ? 1 : 0
+  name        = "${var.customer_name}-${var.environment}-edxapp-course-discovery-s3-full-access"
+  description = "Grants full access to s3 resources in the ${aws_s3_bucket.edxapp_course_discovery.id} bucket"
+
+  policy = data.aws_iam_policy_document.edxapp_course_discovery_s3_full_access.json
```
Comment (giovannicimolin) [inline modules/openedx/s3/course_discovery.tf]:
```suggestion
  name        = "${var.client_shortname}-${var.environment}-edxapp-course-discovery-s3-full-access"
  description = "Grants full access to s3 resources in the ${aws_s3_bucket.edxapp_course_discovery[0].id} bucket"

  policy = data.aws_iam_policy_document.edxapp_course_discovery_s3_full_access[0].json
```

File: modules/openedx/s3/course_discovery.tf [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+resource aws_iam_user edxapp_course_discovery_s3_user {
+  count = var.enable_course_discovery ? 1 : 0
+  name = "${var.customer_name}-${var.environment}-edxapp-course-discovery-s3"
+}
+
+resource aws_iam_user_policy_attachment edxapp_course_discovery_s3_full_access {
+  count = var.enable_course_discovery ? 1 : 0
+  policy_arn = aws_iam_policy.edxapp_course_discovery_s3_full_access.arn
+  user = aws_iam_user.edxapp_course_discovery_s3_user.name
```
Comment (giovannicimolin) [inline modules/openedx/s3/course_discovery.tf]:
```suggestion
  policy_arn = aws_iam_policy.edxapp_course_discovery_s3_full_access[0].arn
  user = aws_iam_user.edxapp_course_discovery_s3_user[0].name
```

File: modules/openedx/s3/course_discovery.tf [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+resource aws_iam_access_key edxapp_course_discovery_s3_access_key {
+  count = var.enable_course_discovery ? 1 : 0
+  user = aws_iam_user.edxapp_course_discovery_s3_user.name
```
Comment (giovannicimolin) [inline modules/openedx/s3/course_discovery.tf]:
```suggestion
  user = aws_iam_user.edxapp_course_discovery_s3_user[0].name
```

**Review by giovannicimolin — CHANGES_REQUESTED** [review]
I've had to make a few changes to get this working (tested on NELC environment).

Also, I think that since we're using the S3 module, it makes sense to split the multiple resource creations in separate files (using the same structure as the other buckets in this folder).

File: modules/openedx/application/outputs.tf (hunk 1 above)
Comment (farhaanbukhsh) [reply modules/openedx/application/outputs.tf]: Ahh okay, is there a way I can test such changes?

File: modules/openedx/s3/course_discovery.tf (hunk 2 above)
Comment (farhaanbukhsh) [reply modules/openedx/s3/course_discovery.tf]: My bad

File: modules/openedx/s3/course_discovery.tf (hunk 3 above)
Comment (farhaanbukhsh) [reply modules/openedx/s3/course_discovery.tf]: Makes sense

File: modules/openedx/application/outputs.tf (hunk 1 above)
Comment (farhaanbukhsh) [reply modules/openedx/application/outputs.tf]: wouldn't this create a issue if `enable_course_discovery` is false and this resource doesn't exist?

Comment (farhaanbukhsh) [reply modules/openedx/application/outputs.tf]: https://discuss.hashicorp.com/t/is-it-possible-to-display-output-conditionally/2155

**Review by giovannicimolin — APPROVED** [review]
:+1:

- [x] I tested this:
1. Using NELP's environment, I've checked out this branch common repo.
2. Tested with `terraform plan`, output:
```
Terraform will perform the following actions:

  # module.application.module.s3.data.aws_iam_policy_document.edxapp_course_discovery_s3_full_access[0] will be read during apply
  # (config refers to values not yet known)
 <= data "aws_iam_policy_document" "edxapp_course_discovery_s3_full_access"  {
      + id      = (known after apply)
      + json    = (known after apply)
      + version = "2012-10-17"

      + statement {
          + actions   = [
              + "s3:*",
            ]
          + effect    = "Allow"
          + resources = [
              + (known after apply),
              + (known after apply),
            ]
        }
    }

  # module.application.module.s3.aws_iam_access_key.edxapp_course_discovery_s3_access_key[0] will be created
  + resource "aws_iam_access_key" "edxapp_course_discovery_s3_access_key" {
      + encrypted_secret     = (known after apply)
      + id                   = (known after apply)
      + key_fingerprint      = (known after apply)
      + secret               = (sensitive value)
      + ses_smtp_password_v4 = (sensitive value)
      + status               = (known after apply)
      + user                 = "nelp-edxapp-production-course-discovery-iam"
    }

  # module.application.module.s3.aws_iam_policy.edxapp_course_discovery_s3_full_access[0] will be created
  + resource "aws_iam_policy" "edxapp_course_discovery_s3_full_access" {
      + arn         = (known after apply)
      + description = (known after apply)
      + id          = (known after apply)
      + name        = "nelp-edxapp-production-course-discovery-s3-full-access"
      + path        = "/"
      + policy      = (known after apply)
    }

  # module.application.module.s3.aws_iam_user.edxapp_course_discovery_s3_user[0] will be created
  + resource "aws_iam_user" "edxapp_course_discovery_s3_user" {
      + arn           = (known after apply)
      + force_destroy = false
      + id            = (known after apply)
      + name          = "nelp-edxapp-production-course-discovery-iam"
      + path          = "/"
      + unique_id     = (known after apply)
    }

  # module.application.module.s3.aws_iam_user_policy_attachment.edxapp_course_discovery_s3_full_access[0] will be created
  + resource "aws_iam_user_policy_attachment" "edxapp_course_discovery_s3_full_access" {
      + id         = (known after apply)
      + policy_arn = (known after apply)
      + user       = "nelp-edxapp-production-course-discovery-iam"
    }

  # module.application.module.s3.aws_s3_bucket.edxapp_course_discovery[0] will be created
  + resource "aws_s3_bucket" "edxapp_course_discovery" {
      + acceleration_status         = (known after apply)
      + acl                         = "private"
      + arn                         = (known after apply)
      + bucket                      = "nelp-edxapp-production-course-discovery"
      + bucket_domain_name          = (known after apply)
      + bucket_regional_domain_name = (known after apply)
      + force_destroy               = false
      + hosted_zone_id              = (known after apply)
      + id                          = (known after apply)
      + region                      = (known after apply)
      + request_payer               = (known after apply)
      + website_domain              = (known after apply)
      + website_endpoint            = (known after apply)

      + cors_rule {
          + allowed_headers = [
              + "*",
            ]
          + allowed_methods = [
              + "PUT",
              + "GET",
            ]
          + allowed_origins = [
              + "*",
            ]
        }

      + versioning {
          + enabled    = (known after apply)
          + mfa_delete = (known after apply)
        }
    }
```
- [x] I read through the code
- [ ] ~~I checked for accessibility issues~~
- [ ] Includes documentation: no. :x:


@farhaanbukhsh This is missing some documentation, but no need to include it in the current ticket, I'll address it once I'm working on the follow-up implementation next sprint.

Can you merge this PR and create a new release with a description of the changes?

**Comment by lgp171188** [comment]
@farhaanbukhsh, @giovannicimolin, for some reason, the changes made in this PR and possibly in the related NELP PR don't print the IAM credentials in the output at all. Do you have any idea what the issue could be? Farhaan and I have been debugging this for a while now and tried out many things. But we are unable to fix it.

**Pull request merged**